alloptions-limit.php

```diff
@@ -39,38 +41,42 @@ function wpcom_vip_sanity_check_alloptions() {
 		wp_cache_add( 'alloptions_size', $alloptions_size, '', 60 );
 	}
 
-	$blocked = $alloptions_size > $alloptions_size_die;
-	$warning = $alloptions_size > $alloptions_size_warn;
+	$warning        = $alloptions_size > $alloptions_size_warn;
+	$maybe_blocked  = $alloptions_size > $alloptions_size_die;
+	$really_blocked = false;
 
-	// If it's at least over the warning threshold (will also run when blocked), notify
-	if ( $warning ) {
-		if ( $blocked ) {
-			add_filter( 'alloptions_overrule_ack', '__return_true' );
-		}
-		// NOTE - This function has built-in rate limiting so it's ok to call on every request
-		wpcom_vip_sanity_check_alloptions_notify( $alloptions_size, $blocked );
+	$alloptions_size_compressed = 0;
+
+	if ( ! $warning ) {
+		return;
 	}
 
-	// Will exit with a 503
-	if ( $blocked ) {
-		wpcom_vip_sanity_check_alloptions_die( $alloptions );
+	if ( $maybe_blocked ) {
+		// It's likely at this point the site is already experiencing performance degradation.
+		// We're using gzdeflate here because pecl-memcache uses Zlib compression for large values.
+		// See https://github.com/websupport-sk/pecl-memcache/blob/e014963c1360d764e3678e91fb73d03fc64458f7/src/memcache_pool.c#L303-L354
+		$alloptions_size_compressed = wp_cache_get( 'alloptions_size_compressed' );
+		if ( ! $alloptions_size_compressed ) {
+			// phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
+			$alloptions_size_compressed = strlen( gzdeflate( serialize( wp_load_alloptions() ) ) );
+			wp_cache_add( 'alloptions_size_compressed', $alloptions_size_compressed, '', 60 );
```

Comment:
Could use `maybe_serialize()` here. While `wp_load_alloptions()` _should_ give an array, it is filterable and doesn't type check before the return.

Similarly,  `gzdeflate` will return false if it fails, so may need to handle that.